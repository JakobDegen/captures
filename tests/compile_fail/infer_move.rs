use captures::*;

fn takes_static<T: 'static + FnOnce() -> i32>(f: T) -> i32 {
    f()
}

// Dont emit a move if we dont need to
fn main() {
    let mut a: i32 = 0;
    let _ = takes_static(capture!(
        all a,
        || { a += 2; a }
    ));
}
