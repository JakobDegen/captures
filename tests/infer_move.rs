use captures::*;

fn takes_static<T: 'static + FnOnce() -> i32>(f: T) -> i32 {
    f()
}

// Checks that we correctly make a closure taking `clone` move
fn clone_dir() {
    let a: i32 = 0;
    let out = takes_static(capture!(
        clone mut a,
        || { a += 2; a }
    ));
    assert_eq!(out, 2);
}

// Checks that we correcty make a closure taking `with` move
fn with_dir() {
    let out = takes_static(capture!(
        with mut a = 10,
        || { a += 2; a }
    ));
    assert_eq!(out, 12);
}

// Checks that we don't make mistakes when combining lots of these
fn combination() {
    let a: i32 = 0;
    let b: i32 = 3;
    let out = takes_static(capture!(
        clone a,
        clone b,
        with mut c = 10,
        with d = 100,
        || { c += a + b + d; c }
    ));
    assert_eq!(out, 113);
}

// Dont emit garbage if the user provides one
fn unnecessary() {
    let out = takes_static(capture!(
        with d = 50,
        move || d + 10
    ));
    assert_eq!(out, 60);
}

fn main() {
    clone_dir();
    with_dir();
    combination();
    unnecessary();
}
