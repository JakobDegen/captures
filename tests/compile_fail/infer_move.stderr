error[E0373]: closure may outlive the current function, but it borrows `a`, which is owned by the current function
  --> tests/compile_fail/infer_move.rs:12:9
   |
12 |         || { a += 2; a }
   |         ^^   - `a` is borrowed here
   |         |
   |         may outlive borrowed value `a`
   |
note: function requires argument type to outlive `'static`
  --> tests/compile_fail/infer_move.rs:10:13
   |
10 |       let _ = takes_static(capture!(
   |  _____________^
11 | |         all a,
12 | |         || { a += 2; a }
13 | |     ));
   | |______^
help: to force the closure to take ownership of `a` (and any other referenced variables), use the `move` keyword
   |
12 |         move || { a += 2; a }
   |         ++++
